flash/core/data/data_source.py

```diff
@@ -352,6 +352,65 @@ def generate_dataset(
             return dataset
 
 
+class DataSourceCollection(dict):
+    def __init__(self, data_sources: Dict[str, DataSource], default_data_source: str):
+        super().__init__(**data_sources)
+        self.default_data_source = default_data_source
+        self[None] = self[default_data_source]
+        self["default"] = self[default_data_source]
+
+        self._train_ds: Optional[BaseAutoDataset] = None
+        self._val_ds: Optional[BaseAutoDataset] = None
+        self._test_ds: Optional[BaseAutoDataset] = None
+        self._predict_ds: Optional[BaseAutoDataset] = None
+
+        self._data_source: Optional[DataSource] = None
+
+    def _execute(self, data_source_enum: DefaultDataSources, *args, **kwargs):
+        if data_source_enum in self:
+            func_name = f"from_{data_source_enum.name.lower()}"
+            self._data_source = getattr(self[data_source_enum], func_name)
+            self._train_ds, self._val_ds, self._test_ds, self._predict_ds = self._data_source(*args, **kwargs)
+            return
+        raise NotImplementedError
+
+    def from_folders(self, *args, **kwargs):
+        self._execute(DefaultDataSources.FOLDERS, *args, **kwargs)
+
+    def from_files(self, *args, **kwargs):
+        self._execute(DefaultDataSources.FILES, *args, **kwargs)
+
+    def from_numpy(self, *args, **kwargs):
+        self._execute(DefaultDataSources.NUMPY, *args, **kwargs)
+
+    def from_tensors(self, *args, **kwargs):
+        self._execute(DefaultDataSources.TENSORS, *args, **kwargs)
+
+    def from_csv(self, *args, **kwargs):
+        self._execute(DefaultDataSources.CSV, *args, **kwargs)
+
+    def from_json(self, *args, **kwargs):
+        self._execute(DefaultDataSources.JSON, *args, **kwargs)
+
+    def from_datasets(self, *args, **kwargs):
+        self._execute(DefaultDataSources.DATASETS, *args, **kwargs)
+
+    def from_fiftyone(self, *args, **kwargs):
+        self._execute(DefaultDataSources.FIFTYONE, *args, **kwargs)
+
+    def from_dataframe(self, *args, **kwargs):
+        self._execute(DefaultDataSources.DATAFRAME, *args, **kwargs)
+
+    def from_lists(self, *args, **kwargs):
+        self._execute(DefaultDataSources.LISTS, *args, **kwargs)
+
+    def from_sentences(self, *args, **kwargs):
```

Comment:
Not a fan of having application-specific code here, a Registry style implementation might work better and reduce code surface. All these functions are basically just calling the same function with a different key, which sounds like a lookup to me rather than hardcoding all of these.

Reply:
Yes, I know it is not ideal, but this is primarly for IDE resolution, so contributors can follow the function call and understand where the code goes.

Reply:
Yeah that's entirely fair, it just doesn't seem very scalable because now wouldn't we have to implement a custom "from" call both here and in the data source itself? For instance, I have a "from_config_file" method on a derived DataSource class which parses a yaml and loads up my data module.

I'm thinking it might be preferable in the longrun to abstract these away into a more scalable API? I can imaging tons of different "from_*" calls ending up making a lot of sense for different use cases all with different logic.